python-interface.cpp

```diff
@@ -0,0 +1,264 @@
+    {"getChannelName", PythonInterface::getChannelName, METH_VARARGS, "Get the irc channel name"},
+    {NULL, NULL, 0, NULL} // Sentinel value, end of Python API methods.
+};
+
+/* Define the new Python module */
+static struct PyModuleDef ircmodule = {
+    PyModuleDef_HEAD_INIT,
+    "ircbot",  // name
+    NULL,      // no docstring
+    -1,        // size (-1 means module has global state)
+    IrcMethods // module methods
+};
+
+/* Create the new Python module */
+PyMODINIT_FUNC PyInitIrc(void)
+{
+    return PyModule_Create(&ircmodule);
+}
+
+PyObject * PythonInterface::sendPythonMessage(PyObject *self, PyObject *args)
+{
+  const char * message;
+  if(!PyArg_ParseTuple(args, "s", &message))
+      Py_RETURN_FALSE;
+  string cMessage = message;
+  ircbot->sendMessage(cMessage);
+  Py_RETURN_TRUE;
+}
+
+PyObject * PythonInterface::sendPythonChannelMessage(PyObject *self, PyObject *args)
+{
+  string msg = (*ircbot).getChannelSendString();
+  const char * message;
+  if(!PyArg_ParseTuple(args, "s", &message))
+      Py_RETURN_FALSE;
+  string messageOnly = message;
+  msg  += messageOnly;
+  // add carridge return and new line to the end of messages
+  msg += "\r\n";
+  // update the history file
+  ircbot->writeHistory(ircbot->getNick(), "", messageOnly + "\r\n");
+  send(ircbot->connectionSocket,msg.c_str(),msg.length(),0);
+  Py_RETURN_TRUE;
+}
+
+PyObject * PythonInterface::runSystemCommand(PyObject *self, PyObject *args)
+{
+  const char * command;
+  if(!PyArg_ParseTuple(args, "s", &command))
+      Py_RETURN_FALSE;
+  system(command);
+  Py_RETURN_TRUE;
+}
+
+PyObject * PythonInterface::runSystemCommandWithOutput(PyObject *self, PyObject *args)
+{
+  const char * command;
+  if(!PyArg_ParseTuple(args, "s", &command))
+      Py_RETURN_TRUE;
+  string ret =  systemUtils->runProcessWithReturn(command);
+  return PyBytes_FromString(ret.c_str());
+}
+
+PyObject * PythonInterface::getBotName(PyObject *self, PyObject *args)
+{
+  return PyBytes_FromString(ircbot->getNick().c_str());
+}
+
+PyObject * PythonInterface::getChannelName(PyObject *self, PyObject *args)
+{
+  return PyBytes_FromString(ircbot->getChannel().c_str());
+}
+
+PyObject * PythonInterface::sendPythonPrivateMessage(PyObject *self, PyObject *args)
+{
+  const char *username, *message;
+  if(!PyArg_ParseTuple(args, "ss", &username, &message))
+      Py_RETURN_FALSE;
+  ircbot->outputToUser(username, message);
+  Py_RETURN_TRUE;
+}
+
+
+PyObject * PythonInterface::sendPythonMessageToSource(PyObject *self, PyObject *args)
+{
+  const char *username, *message, *isPrivate;
+
+  if(!PyArg_ParseTuple(args, "sss", &username, &message, &isPrivate)) {
+      PyErr_Print();
+      Py_RETURN_FALSE;
+  }
+
+  // TODO: Should be able to parse it as a char or actually an int/bool and skip this conversion
+  bool isPrivateMessage = (*isPrivate != '0');
+
+  string msg = ircbot->getChannelSendString();
+  msg  += message;
+  msg += "\r\n";
```

Comment:
Yeah, very true. Python scripting implementation copied some of the bad habits just to get it working. The whole interface(s) need a tidy up at some point.